File: apps/cli/clients/python/src/dokugen/commands/og.py
import os
import sys
import json
import base64
import requests
from rich.console import Console
from dokugen import utils
from dokugen.project_detect import detect_project_type

console = Console()

DOKUGEN_BANNER = """
   ___   ____  __ ____  _____________  __
  / _ \\ / __ \\/ //_/ / / / ____/ ____/ / /
 / / / / / / / ,< / / / / / __/ __/ / /_ 
/ /_/ / /_/ / /| / /_/ / /_/ / /___/ / / 
\\____/\\____/_/ |_\\____/\\____/_____/_/ /  
                                   /_/   
"""


def register_og_parser(subparsers):
    parser = subparsers.add_parser(
        "og",
        help="Generate a beautiful 1200x630 OG social preview card for your project",
    )
    parser.add_argument(
        "--force-new",
        action="store_true",
        help="Force recreate the .dokugen/card.json configuration file using AI",
    )


def cmd_og(args):
    if not utils.is_git_repository():
        console.print("[red]No Git repository found. Please navigate to a project directory that has a Git repository.[/red]")
        sys.exit(1)

    utils.check_and_update()
    console.print(DOKUGEN_BANNER, style="#000080")

    project_dir = os.getcwd()
    project_name = os.path.basename(project_dir)
    dokugen_dir = os.path.join(project_dir, ".dokugen")
    os.makedirs(dokugen_dir, exist_ok=True)

    config_path = os.path.join(dokugen_dir, "card.json")
    png_path = os.path.join(dokugen_dir, "card.png")
    seo_path = os.path.join(dokugen_dir, "seo-instructions.txt")

    has_config = os.path.exists(config_path)
    force_new = getattr(args, "force_new", False)

    with utils.create_spinner("Checking internet..."):
        has_internet = utils.check_internet_connection()

    if not has_internet:
        console.print("[red]Please check your internet connection and try again.[/red]")
        return

    metadata = None

    if not has_config or force_new:
        with utils.create_ticking_spinner("Analyzing project to generate card profile..."):
            try:
                project_type = detect_project_type(project_dir)
                codebase_summary = f"Project Name: {project_name}\nDetected Tech Stack: {project_type}\n"

                readme_path = os.path.join(project_dir, "README.md")
                if os.path.exists(readme_path):
                    with open(readme_path, "r", encoding="utf-8", errors="ignore") as f:
                        readme_content = f.read()
                    codebase_summary += f"\nExisting README Summary:\n{readme_content[:3000]}"
                else:
                    try:
                        files = os.listdir(project_dir)
                        src_files = [f for f in files if f.endswith((".ts", ".js", ".py", ".go", ".rs", ".cpp", ".h", ".java"))]
                        if src_files:
                            codebase_summary += f"\nCore source files found: {', '.join(src_files)}"
                    except Exception:
                        pass

                css_candidates = [
                    "index.css", "globals.css", "src/index.css", "src/globals.css",
                    "app/globals.css", "tailwind.config.js", "tailwind.config.ts"
                ]
                for css_file in css_candidates:
                    full_css_path = os.path.join(project_dir, css_file)
                    if os.path.exists(full_css_path):
                        with open(full_css_path, "r", encoding="utf-8", errors="ignore") as cf:
                            css_content = cf.read()
                        codebase_summary += f"\nBrand CSS Styling ({css_file}):\n{css_content[:1500]}"
                        break

                backend_url = utils.get_backend_domain()
                res = requests.post(
                    f"{backend_url}/api/og-metadata",
                    json={"summary": codebase_summary},
                    timeout=30,
                )

                if res.status_code != 200:
                    console.print(f"[red]Failed to generate metadata profile via server: {res.text}[/red]")
                    return

                metadata = res.json()
                with open(config_path, "w", encoding="utf-8") as f:
                    json.dump(metadata, f, indent=2)

                console.print("[green]✔ Created configuration: .dokugen/card.json[/green]")

            except Exception as e:
                console.print(f"[red]Failed to generate card profile: {e}[/red]")
                return
    else:
        with open(config_path, "r", encoding="utf-8") as f:
            metadata = json.load(f)

    with utils.create_ticking_spinner("Rendering card PNG..."):
        try:
            logo_path = metadata.get("logo", "")
            if logo_path and not logo_path.startswith("http") and not logo_path.startswith("data:"):
                local_logo = os.path.abspath(os.path.join(project_dir, logo_path))
                if os.path.exists(local_logo):
                    with open(local_logo, "rb") as lf:
                        encoded = base64.b64encode(lf.read()).decode("utf-8")
                    mime = "image/svg+xml" if local_logo.endswith(".svg") else "image/png"
                    metadata["logo"] = f"data:{mime};base64,{encoded}"

            backend_url = utils.get_backend_domain()
            res = requests.post(
                f"{backend_url}/api/render-og",
                json=metadata,
                timeout=20,
            )

            if res.status_code != 200:
                console.print(f"[red]Failed to render card image: {res.text}[/red]")
                return

            png_bytes = res.content
            with open(png_path, "wb") as f:
                f.write(png_bytes)

            seo_text = f"""<!-- Open Graph Meta Tags (Copy & Paste into your HTML <head> or Next.js metadata) -->
<meta property="og:title" content="{metadata.get('title', '')}" />
{f'<meta property="og:description" content="{metadata.get("tagline", "")}" />' if metadata.get("tagline") else ''}
<meta property="og:image" content="https://yourdomain.com/.dokugen/card.png" />
<meta property="og:image:width" content="1200" />
<meta property="og:image:height" content="630" />

<!-- Twitter Card Meta Tags -->
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="{metadata.get('title', '')}" />
{f'<meta name="twitter:description" content="{metadata.get("tagline", "")}" />' if metadata.get("tagline") else ''}
<meta name="twitter:image" content="https://yourdomain.com/.dokugen/card.png" />"""

            with open(seo_path, "w", encoding="utf-8") as sf:
                sf.write(seo_text)

        except Exception as e:
            console.print(f"[red]Failed to render social card PNG: {e}[/red]")
            return

    console.print("\n[green]✔ Created card image: ./.dokugen/card.png[/green]")
    console.print("[green]✔ Created SEO meta tags: ./.dokugen/seo-instructions.txt[/green]")

File: apps/cli/clients/typescript/bin/commands/og.ts
import { Command } from "commander";
import * as path from "path";
import fs from "fs-extra";
import axios from "axios";
import chalk from "chalk";
import { createSpinner } from "nanospinner";
import { checkAndUpdate, checkInternetConnection, getBackendDomain } from "../helpers/network.js";
import { isGitRepository } from "../helpers/git.js";
import { DOKUGEN_BANNER } from "../helpers/constants.js";
//@ts-ignore
import { detectProjectType } from "../projectDetect.mjs";

interface OgButton {
  label: string;
  variant?: "primary" | "secondary";
}

interface OgMetadata {
  title: string;
  tagline?: string;
  techStack?: string[];
  theme: string;
  url?: string;
  author?: string;
  version?: string;
  logo?: string;
  buttons?: OgButton[];
}

export function registerOgCommand(program: Command) {
  const projectName = path.basename(process.cwd());

  program
    .command("og")
    .description(`Generate a beautiful 1200x630 OG social preview card for ${projectName}`)
    .option("--force-new", "Force recreate the .dokugen/card.json configuration file using AI")
    .action(async (options: any) => {
      if (!isGitRepository()) {
        console.log(
          chalk.red(
            "No Git repository found. Please navigate to a project directory that has a Git repository."
          )
        );
        process.exit(1);
      }

      await checkAndUpdate();
      console.log("\n" + chalk.hex("#000080")(DOKUGEN_BANNER) + "\n");
      const projectDir = process.cwd();
      const dokugenFolder = path.join(projectDir, ".dokugen");
      await fs.ensureDir(dokugenFolder);

      const configPath = path.join(dokugenFolder, "card.json");
      const pngPath = path.join(dokugenFolder, "card.png");
      const seoPath = path.join(dokugenFolder, "seo-instructions.txt");

      const hasConfig = await fs.pathExists(configPath);

      const connectionSpinner = createSpinner("Checking internet...").start();
      const hasInternet = await checkInternetConnection();
      connectionSpinner.stop();

      if (!hasInternet) {
        return console.log(
          chalk.red("Please check your internet connection and try again.")
        );
      }

      let metadata: OgMetadata;

      if (!hasConfig || options.forceNew) {
        const startTime = Date.now();
        const metadataSpinner = createSpinner("Analyzing project to generate card profile...").start();
        const timerInterval = setInterval(() => {
          const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
          metadataSpinner.update({ text: `Analyzing project to generate card profile... (${elapsed}s)` });
        }, 100);

        try {
          const projectType = await detectProjectType(projectDir);
          
          let codebaseSummary = `Project Name: ${projectName}\nDetected Tech Stack: ${projectType}\n`;

          const readmePath = path.join(projectDir, "README.md");
          if (await fs.pathExists(readmePath)) {
            const readmeContent = await fs.readFile(readmePath, "utf-8");
            codebaseSummary += `\nExisting README Summary:\n${readmeContent.slice(0, 3000)}`;
          } else {
            const files = await fs.readdir(projectDir).catch(() => []);
            const srcFiles = files.filter(f => f.match(/\.(ts|js|py|go|rs|cpp|h|java)$/i));
            if (srcFiles.length > 0) {
              codebaseSummary += `\nCore source files found: ${srcFiles.join(", ")}`;
            }
          }

          const cssCandidates = [
            "index.css", "globals.css", "src/index.css", "src/globals.css",
            "app/globals.css", "tailwind.config.js", "tailwind.config.ts"
          ];
          for (const cssFile of cssCandidates) {
            const fullCssPath = path.join(projectDir, cssFile);
            if (await fs.pathExists(fullCssPath)) {
              const cssContent = await fs.readFile(fullCssPath, "utf-8");
              codebaseSummary += `\nBrand CSS Styling (${cssFile}):\n${cssContent.slice(0, 1500)}`;
              break;
            }
          }

          const backendUrl = await getBackendDomain();
          const response = await axios.post<OgMetadata>(`${backendUrl}/api/og-metadata`, {
            summary: codebaseSummary
          }, { timeout: 30000 });

          clearInterval(timerInterval);
          metadataSpinner.stop();

          metadata = response.data;
          await fs.writeJson(configPath, metadata, { spaces: 2 });

          console.log(chalk.green("✔ Created configuration: .dokugen/card.json"));
        } catch (err: any) {
          clearInterval(timerInterval);
          metadataSpinner.error({ text: "Failed to generate card profile." });
          console.error(chalk.red(err.message));
          return;
        }
      } else {
        metadata = await fs.readJson(configPath) as OgMetadata;
      }

      // Automatically render the PNG card image right after
      const renderStartTime = Date.now();
      const renderSpinner = createSpinner("Rendering card PNG...").start();
      const renderTimerInterval = setInterval(() => {
        const elapsed = ((Date.now() - renderStartTime) / 1000).toFixed(1);
        renderSpinner.update({ text: `Rendering card PNG... (${elapsed}s)` });
      }, 100);

      try {
        let logoData = metadata.logo || "";
        if (logoData && !logoData.startsWith("http") && !logoData.startsWith("data:")) {
          const localLogoPath = path.resolve(projectDir, logoData);
          if (await fs.pathExists(localLogoPath)) {
            const logoBuffer = await fs.readFile(localLogoPath);
            const ext = path.extname(localLogoPath).toLowerCase();
            const mimeType = ext === ".svg" ? "image/svg+xml" : "image/png";
            logoData = `data:${mimeType};base64,${logoBuffer.toString("base64")}`;
          }
        }

        const renderPayload = { ...metadata, logo: logoData };

        const backendUrl = await getBackendDomain();
        const response = await axios.post(`${backendUrl}/api/render-og`, renderPayload, {
          responseType: "arraybuffer",
          timeout: 20000
        });

        clearInterval(renderTimerInterval);
        renderSpinner.stop();

        await fs.writeFile(pngPath, Buffer.from(response.data as any));

        const seoText = `
<!-- Open Graph Meta Tags (Copy & Paste into your HTML <head> or Next.js metadata) -->
<meta property="og:title" content="${metadata.title}" />
${metadata.tagline ? `<meta property="og:description" content="${metadata.tagline}" />\n` : ""}<meta property="og:image" content="https://yourdomain.com/.dokugen/card.png" />
<meta property="og:image:width" content="1200" />
<meta property="og:image:height" content="630" />

<!-- Twitter Card Meta Tags -->
<meta name="twitter:card" content="summary_large_image" />
<meta name="twitter:title" content="${metadata.title}" />
${metadata.tagline ? `<meta name="twitter:description" content="${metadata.tagline}" />\n` : ""}<meta name="twitter:image" content="https://yourdomain.com/.dokugen/card.png" />
`.trim();

        await fs.writeFile(seoPath, seoText);

        console.log(chalk.green(`\n✔ Created card image: ./.dokugen/card.png`));
        console.log(chalk.green(`✔ Created SEO meta tags: ./.dokugen/seo-instructions.txt`));
      } catch (err: any) {
        clearInterval(renderTimerInterval);
        renderSpinner.error({ text: "Failed to render social card PNG." });
        console.error(chalk.red(err.message));
      }
    });
}
